presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java

```diff
@@ -0,0 +1,455 @@
+                    }
+                    if (!columnFamily.getCompression().toString().equals("NONE")) {
+                        properties.put(PhoenixTableProperties.COMPRESSION, columnFamily.getCompression().toString());
+                    }
+                    if (columnFamily.getTimeToLive() < FOREVER) {
+                        properties.put(PhoenixTableProperties.TTL, columnFamily.getTimeToLive());
+                    }
+                    break;
+                }
+            }
+        }
+        catch (IOException | SQLException e) {
+            throw new PrestoException(PHOENIX_METADATA_ERROR, "Couldn't get Phoenix table properties", e);
+        }
+        return properties.build();
+    }
+
+    @Override
+    public void createSchema(ConnectorSession session, String schemaName, Map<String, Object> properties)
+    {
+        checkArgument(!DEFAULT_SCHEMA.equalsIgnoreCase(schemaName), "Can't create 'default' schema which maps to Phoenix empty schema");
+        checkArgument(properties.isEmpty(), "Can't have properties for schema creation");
+        phoenixClient.execute(session, format("CREATE SCHEMA %s", getEscapedArgument(toMetadataCasing(session, schemaName))));
+    }
+
+    @Override
+    public void dropSchema(ConnectorSession session, String schemaName)
+    {
+        checkArgument(!DEFAULT_SCHEMA.equalsIgnoreCase(schemaName), "Can't delete 'default' schema which maps to Phoenix empty schema");
+        phoenixClient.execute(session, format("DROP SCHEMA %s", getEscapedArgument(toMetadataCasing(session, schemaName))));
+    }
+
+    private String toMetadataCasing(ConnectorSession session, String schemaName)
+    {
+        try (Connection connection = phoenixClient.getConnection(JdbcIdentity.from(session))) {
+            boolean uppercase = connection.getMetaData().storesUpperCaseIdentifiers();
+            if (uppercase) {
+                schemaName = schemaName.toUpperCase(ENGLISH);
+            }
+        }
+        catch (SQLException e) {
+            throw new PrestoException(PHOENIX_METADATA_ERROR, "Couldn't get casing for the schema name", e);
+        }
+        return schemaName;
+    }
+
+    @Override
+    public void createTable(ConnectorSession session, ConnectorTableMetadata tableMetadata, boolean ignoreExisting)
+    {
+        JdbcTableHandle existingTable = getTableHandle(session, tableMetadata.getTable());
+        if (existingTable != null && !ignoreExisting) {
+            throw new IllegalArgumentException("Target table already exists: " + tableMetadata.getTable());
+        }
+        createTable(session, tableMetadata);
+    }
+
+    @Override
+    public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, ConnectorTableMetadata tableMetadata, Optional<ConnectorNewTableLayout> layout)
+    {
+        JdbcOutputTableHandle handle = createTable(session, tableMetadata);
```

Comment:
The table creation needs to happen in `finishCreateTable`. The reason is that `beginCreateTable` is called during query planning, so an `EXPLAIN CREATE TABLE xxx AS ...` or similar would end up creating the table.

Reply:
Not sure I follow here.  My understanding is `beginCreateTable` is used to create a temporary table, then data is inserted, then `finishCreateTable` is used to rename the table so the entire operation is atomic.  So don't we need to create the table in `beginCreateTable` ?  
Phoenix doesn't support renaming of tables, so `finishCreateTable` here doesn't do anything.  Other connectors I looked at do the same (Accumulo, Cassandra, Kudu).

Reply:
Your understanding is correct. And you're right, I think this is the best we can do for now. We need to fix this in the connector API so that creation of the handle during planning (needed since it has to go into the query plan) is separate from the start of execution.